From 484d44fea36ca8d14a67e536be054d3371df5214 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:20:13 -0400
Subject: [net] netfilter: nft_rbtree: no need for spinlock from set destroy path

Message-id: <20141021162013.22888.93547.stgit@localhost.localdomain>
Patchwork-id: 97994
O-Subject: [RHEL7 net PATCH 097/124] netfilter: nft_rbtree: no need for spinlock from set destroy path
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: d99407f42f05843ae9e23696ea6d91529d9600db

 commit d99407f42f05843ae9e23696ea6d91529d9600db
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Tue Sep 2 11:29:56 2014 +0200

    netfilter: nft_rbtree: no need for spinlock from set destroy path

    The sets are released from the rcu callback, after the rule is removed
    from the chain list, which implies that nfnetlink cannot update the
    rbtree and no packets are walking on the set anymore. Thus, we can get
    rid of the spinlock in the set destroy path there.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>
    Reviewied-by: Thomas Graf <tgraf@suug.ch>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nft_rbtree.c b/net/netfilter/nft_rbtree.c
index e1836ff..46214f2 100644
--- a/net/netfilter/nft_rbtree.c
+++ b/net/netfilter/nft_rbtree.c
@@ -234,13 +234,11 @@ static void nft_rbtree_destroy(const struct nft_set *set)
  struct nft_rbtree_elem *rbe;
  struct rb_node *node;
 
- spin_lock_bh(&nft_rbtree_lock);
  while ((node = priv->root.rb_node) != NULL) {
   rb_erase(node, &priv->root);
   rbe = rb_entry(node, struct nft_rbtree_elem, node);
   nft_rbtree_elem_destroy(set, rbe);
  }
- spin_unlock_bh(&nft_rbtree_lock);
 }
 
 static bool nft_rbtree_estimate(const struct nft_set_desc *desc, u32 features,
-- 
1.7.1